appconfig/appconfig.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2021 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package appconfig loads repository configuration for GitHub apps. It
+// supports loading directly from a file in a repository, loading from remote
+// references, and loading an organization-level default. The config itself can
+// be in any format.
+package appconfig
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/google/go-github/v37/github"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+)
+
+// RemoteRefParser attempts to parse a RemoteRef from bytes. The parser should
+// return nil with a nil error if b does not encode a RemoteRef and nil with a
+// non-nil error if b encodes an invalid RemoteRef.
+type RemoteRefParser func(path string, b []byte) (*RemoteRef, error)
+
+// RemoteRef identifies a configuration file in a different repository.
+type RemoteRef struct {
+	// The repository in "owner/name" format.
+	Remote string `yaml:"remote" json:"remote"`
+
+	// The path to the config file in the repository.
+	Path string `yaml:"path" json:"path"`
+
+	// The reference (branch, tag, or SHA) to read in the repository. If empty,
+	// use the default branch of the repository.
+	Ref string `yaml:"ref" json:"ref"`
+}
+
+func (r RemoteRef) SplitRemote() (owner, repo string, err error) {
+	slash := strings.IndexByte(r.Remote, '/')
+	if slash <= 0 || slash >= len(r.Remote)-1 {
+		return "", "", errors.Errorf("invalid remote value: %s", r.Remote)
+	}
+	return r.Remote[:slash], r.Remote[slash+1:], nil
+}
+
+// Config contains unparsed configuration data and metadata about where it was found.
+type Config struct {
+	Content []byte
+
+	Source   string
+	Path     string
+	IsRemote bool
+}
+
+// IsUndefined returns true if the Config's content is empty and there is no
+// metadata giving a source.
+func (c Config) IsUndefined() bool {
+	return len(c.Content) == 0 && c.Source == "" && c.Path == ""
+}
+
+// Loader loads configuration for repositories.
+type Loader struct {
+	paths []string
+
+	parser       RemoteRefParser
+	defaultRepo  string
+	defaultPaths []string
+}
+
+// NewLoader creates a Loader that loads configuration from paths.
+func NewLoader(paths []string, opts ...Option) *Loader {
+	defaultPaths := make([]string, len(paths))
+	for i, p := range paths {
+		defaultPaths[i] = strings.TrimPrefix(p, ".github/")
```

Comment:
Why does this path get trimmed out?

Reply:
Mostly just default behavior I decided on. The idea is that if you are using a `.github` repository, you should be able to set configuration for the `.github` repository that is different from the configuration it contains for all the other organization repositories. If the path was the same, the configuration you used for all repos would have to be the same as the configuration used for the `.github` repository itself.

This is all customizable with the `WithOwnerDefault` option.

Reply:
I should also add that stripping this path matches the idea of the `.github` repo being like the shared contents of a `.github` folder in a single repo.